### PR TITLE
JobQueue Tuning

### PR DESCRIPTION
**Pull request by mtrippled**
Allow manual configuration of worker threads. For automatic thread pool
sizing, have different rules for validators and non-validators. Increase
thread count for client handlers to better sustain heavy RPC traffic.
Validator sizing remains unchanged.

### CODE REVIEW

File: src/ripple/app/main/Application.cpp [hunk 1]
```diff
@@ -404,7 +404,9 @@ class ApplicationImp
         //
         , m_jobQueue (std::make_unique<JobQueue>(
             m_collectorManager->group ("jobq"), m_nodeStoreScheduler,
-            logs_->journal("JobQueue"), *logs_))
+            logs_->journal("JobQueue"), *logs_,
+            config_->exists (SECTION_VALIDATOR_TOKEN) ||
```
Comment (bachase) [inline src/ripple/app/main/Application.cpp]: I think this is correct, but paging @wilsonianb to be sure it is forward compatible with dynamic UNL.

File: src/ripple/core/JobQueue.h [hunk 2]
```diff
@@ -108,7 +108,7 @@ class JobQueue
     using JobFunction = std::function <void(Job&)>;
 
     JobQueue (beast::insight::Collector::ptr const& collector,
-        Stoppable& parent, beast::Journal journal, Logs& logs);
+        Stoppable& parent, beast::Journal journal, Logs& logs, bool validator);
```
Comment (bachase) [inline src/ripple/core/JobQueue.h]: Nit: It would be nice to pass this to the `setThreadCount` call rather than adding a member that won't be used much. Or set the number of threads in the constructor as in #2136.

File: src/ripple/core/impl/JobQueue.cpp [hunk 3]
```diff
@@ -161,10 +162,18 @@ JobQueue::setThreadCount (int c, bool const standaloneMode)
...
     {
         c = static_cast<int>(std::thread::hardware_concurrency());
-        c = 2 + std::min (c, 4); // I/O will bottleneck
-
-        JLOG(m_journal.info()) << "Auto-tuning to " << c <<
-                            " validation/transaction/proposal threads";
+        if (validator_)
+            c = 2 + std::min(c, 4); // I/O will bottleneck
+        else
+            c *= 2;
```
Comment (bachase) [inline src/ripple/core/impl/JobQueue.cpp]: Add a comment for why `*=2` for the future?

File: src/ripple/core/JobQueue.h (hunk 2 above)
Comment (nbougalis) [inline src/ripple/core/JobQueue.h]: Nitpick: conceptually, I don't like passing this information into `JobQueue`. No change necessary at this time.

File: src/ripple/core/impl/JobQueue.cpp [hunk 4]
```diff
@@ -161,10 +162,18 @@ JobQueue::setThreadCount (int c, bool const standaloneMode)
     else if (c == 0)
     {
         c = static_cast<int>(std::thread::hardware_concurrency());
-        c = 2 + std::min (c, 4); // I/O will bottleneck
-
-        JLOG(m_journal.info()) << "Auto-tuning to " << c <<
-                            " validation/transaction/proposal threads";
+        if (validator_)
+            c = 2 + std::min(c, 4); // I/O will bottleneck
```
Comment (mtrippled) [inline src/ripple/core/impl/JobQueue.cpp]: I would actually have validators and client handlers get the same treatment, but that's controversial. For this PR, I was thinking to not affect validators by default.

File: src/ripple/core/JobQueue.h (hunk 2 above)
Comment (mtrippled) [inline src/ripple/core/JobQueue.h]: fixed

File: doc/rippled-example.cfg [hunk 5]
```diff
@@ -690,6 +690,14 @@
 #
 #
 #
+# [workers]
+#
+#   Configures the number of threads for processing work submitted by peers
+#   and clients. If 0, then formulas are used to automatically determine the
```
Comment (mtrippled) [inline doc/rippled-example.cfg]: fixed

File: src/ripple/core/impl/Config.cpp [hunk 6]
```diff
@@ -413,6 +413,9 @@ void Config::loadFromString (std::string const& fileContents)
     if (getSingleSection (secConfig, SECTION_DEBUG_LOGFILE, strTemp, j_))
         DEBUG_LOGFILE       = strTemp;
 
+    if (getSingleSection (secConfig, SECTION_WORKERS, strTemp, j_))
+        WORKERS      = beast::lexicalCastThrow <std::size_t> (strTemp);
```
Comment (mtrippled) [inline src/ripple/core/impl/Config.cpp]: There are boxes with >64 hyperthreads. This will keep increasing.

File: src/ripple/core/JobQueue.h (hunk 2 above)
Comment (mtrippled) [inline src/ripple/core/JobQueue.h]: fixed

**Comment by bachase** [comment]
👍 on latest diff as well.

File: src/ripple/app/main/Application.cpp (hunk 1 above)
Comment (wilsonianb) [inline src/ripple/app/main/Application.cpp]: :+1:

File: src/ripple/core/impl/Config.cpp (hunk 6 above)
Comment (wilsonianb) [inline src/ripple/core/impl/Config.cpp]: You could have an error check based on some multiple of `std::thread::hardware_concurrency`.

Comment (mtrippled) [inline src/ripple/core/impl/Config.cpp]: That's not necessarily ideal either. The config parameter is also for testing, so should allow for sub-optimal values.

**Comment by nbougalis** [comment]
Suggest rewording commit:

```
Improve automatic tuning of thread pool:

The job queue can automatically tune the number of threads that
it creates based on the number of processors or processor cores
that are available.

The existing tuning was very conservative, limiting the maximum
number of threads to only 6.

Adjust the new algorithm to allow a larger number of threads and
allow server administrators to override the value in the config
file.
```

**Comment by mtrippled** [comment]
@nbougalis I reworded the commit.

**Comment by nbougalis** [comment]
Merged as 68b8ffdb638d07937f841f7217edeb25efdb3b5d
